### PR TITLE
changes for parallel put to S3

### PR DESCRIPTION
**Pull request by d-w-moore**
More-or-less the PRC end of the capability behind the UGM S3 parallel put/get Lightning Talk.

Send numthreads/datasize kw on first open( ) operation in the S3 plugin.

Conform with iput/iget per-thread byte counts

Data objects' open(...) in PRC can now be deferred operations, to simplify the code organization for the auto-switching parallel put.

### CODE REVIEW

**Comment by d-w-moore** [comment]
However, see: https://github.com/irods/irods_resource_plugin_s3/issues/2020

**Review by alanking — COMMENTED** [review]
Some (possibly naive) comments from me. Overall I think it's fine

File: MISC.rst [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+  AWS Access Key ID [None]: minioadmin
+  AWS Secret Access Key [None]: minioadmin
+  Default region name [None]: 
+  Default output format [None]: 
+
+  $ aws --endpoint-url http://127.1:9000 s3 mb s3://irods
+
+Set up s3 credentials for the iRODS s3 storage resource::
+
+  $ sudo su - irods -c "/bin/echo -e 'minioadmin\nminioadmin' >/var/lib/irods/s3-credentials"
```
[the remark this answers is not in the document]
Comment (trel) [reply MISC.rst:33]: that's a newline into a credentials file

[the remark this answers is not in the document]
Comment (alanking) [reply MISC.rst:33]: OH. I got my slash directions mixed up. My b

File: MISC.rst [hunk 2]
```diff
@@ -0,0 +1,63 @@
+=============
```
Comment (trel) [inline MISC.rst:1]: have questions about a 'new' .rst file.     I know the README.rst is already long... but ... unless we have an overall plan, I don't want to start breaking them up or having additional files... it will make that information effectively 'hidden'.

So, consider placing a new section in the existing README for now.   Maybe breaking up the docs is a separate issue/effort.

**Review by trel — COMMENTED** [review]
need more documentation in the code about 'why' and 'reasoning'.   lots of very subtle changes here.

File: irods/manager/data_object_manager.py [hunk 3]
```diff
@@ -126,21 +130,23 @@ def put(self, local_path, irods_path, return_data_object = False, num_threads =
...
             obj = irods_path
 
-        with open(local_path, 'rb') as f, self.open(obj, 'w', **options) as o:
-
-            if self.should_parallelize_transfer (num_threads, f):
+        with open(local_path, 'rb') as f:
+            sizelist = []
+            if self.should_parallelize_transfer (num_threads, f, measured_obj_size = sizelist):
+                o = deferred_call( self.open, (obj, 'w'), options)
                 f.close()
```
[the remark this answers is not in the document]
Comment (d-w-moore) [reply irods/manager/data_object_manager.py:137]: I agree it upsets the programmer's normal sensibilities for balance, but in Python an extra close (unlike `fclose` in C/C++) does not matter.  Python will take it like a trooper....

[the remark this answers is not in the document]
Comment (trel) [reply irods/manager/data_object_manager.py:137]: and... it's official.

https://docs.python.org/3.9/library/io.html#io.IOBase.close

> As a convenience, it is allowed to call this method more than once; only the first call, however, will have an effect.

File: irods/parallel.py [hunk 4]
```diff
@@ -385,17 +423,14 @@ def io_main( session, Data, opr_, fname, R='', **kwopt):
     if Operation.isGet():
         total_bytes = Io.seek(0,os.SEEK_END)
         Io.seek(0,os.SEEK_SET)
-    else:
-        with open(fname, 'rb') as f:
-            f.seek(0,os.SEEK_END)
-            total_bytes = f.tell()
+    else: # isPut
+        if total_bytes < 0:
```
[the remark this answers is not in the document]
Comment (d-w-moore) [reply irods/parallel.py:427]: No, just determining the total size in bytes of the transfer in case it has not been specified by the caller.

File: MISC.rst (hunk 2 above)
Comment (d-w-moore) [reply MISC.rst:1]: Should it be treated as part of the natural flow of the README?  Most of the existing topics are tutorials and guides for usage, and this is more a guide for setup and test of S3/Parallel Transfer.

File: irods/manager/data_object_manager.py [hunk 5]
```diff
@@ -126,21 +130,23 @@ def put(self, local_path, irods_path, return_data_object = False, num_threads =
         else:
             obj = irods_path
 
-        with open(local_path, 'rb') as f, self.open(obj, 'w', **options) as o:
-
-            if self.should_parallelize_transfer (num_threads, f):
+        with open(local_path, 'rb') as f:
+            sizelist = []
+            if self.should_parallelize_transfer (num_threads, f, measured_obj_size = sizelist):
```
Comment (korydraughn) [inline irods/manager/data_object_manager.py:135]: `f` is passed as the `obj_sz`?

reading a bit more of the code and alan's comment, i'm guessing `obj_sz` is treated as a variant-like type?

[the remark this answers is not in the document]
Comment (d-w-moore) [reply irods/manager/data_object_manager.py:135]: That was a lesser-of-two-evils situation.  The logic would've been far more complicated without such a device, and Python offers no better choices for how to implement out-variables.  Would a comment be enough?

File: irods/parallel.py [hunk 6]
```diff
@@ -293,19 +307,27 @@ def _io_part (objHandle, range_, file_, opr_, mgr_, thread_debug_id = '', queueO
...
-    num_threads = max(1, min(multiprocessing.cpu_count(), num_threads))
+    def bytes_range_for_thread( i, num_threads, total_bytes,  chunk ):
+        begin_offs = i * chunk
+        if i + 1 < num_threads:
+            end_offs = (i + 1) * chunk
+        else:
+            end_offs = total_bytes
+        return six.moves.range(begin_offs, end_offs)
+
+    P = total_size // num_threads
```
Comment (korydraughn) [inline irods/parallel.py:324]: is there a word that better describes what `P` represents?

File: irods/parallel.py [hunk 7]
```diff
@@ -293,19 +307,27 @@ def _io_part (objHandle, range_, file_, opr_, mgr_, thread_debug_id = '', queueO
 
 
 def _io_multipart_threaded(operation_ , dataObj_and_IO, replica_token, hier_str, session, fname,
-                           total_size, num_threads = 0, **extra_options):
+                           total_size, num_threads, **extra_options):
     """Called by _io_main.
     Carve up (0,total_size) range into `num_threads` parts and initiate a transfer thread for each one."""
     (D, Io) = dataObj_and_IO
```
Comment (korydraughn) [inline irods/parallel.py:313]: is there a word that better describes what `D` represents?

File: irods/parallel.py [hunk 8]
```diff
@@ -324,7 +346,10 @@ def _io_multipart_threaded(operation_ , dataObj_and_IO, replica_token, hier_str,
...
             Io = session.data_objects.open( D.path, Operation.data_object_mode(initial_open = False),
                                             create = False, finalize_on_close = False,
-                                            **{kw.RESC_HIER_STR_KW: hier_str, kw.REPLICA_TOKEN_KW: replica_token} )
+                                            **{ kw.NUM_THREADS_KW: str(num_threads),
+                                                kw.DATA_SIZE_KW: str(total_size),
+                                                kw.RESC_HIER_STR_KW: hier_str,
+                                                kw.REPLICA_TOKEN_KW: replica_token })
         mgr.add_io( Io )
         if File is None: File = gen_file_handle()
         futures.append(executor.submit( _io_part, Io, r, File, Operation, mgr, str(counter), queueObject))
```
Comment (korydraughn) [inline irods/parallel.py:355]: consider replacing `r` with a word that expresses what it is.

**Review by korydraughn — COMMENTED** [review]
it looks good.

my only suggestion is, prefer words to single letter variables. single letter variables are fine for simple code, but they make it hard to follow the logic in more complex code.

File: irods/manager/data_object_manager.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (korydraughn) [reply irods/manager/data_object_manager.py:135]: java has a similar limitation.

does python have a class that makes it clear a parameter is an out-parameter?

File: MISC.rst [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+
+  $ pip install awscli
+
+  $ aws configure 
+  AWS Access Key ID [None]: minioadmin
+  AWS Secret Access Key [None]: minioadmin
+  Default region name [None]: 
+  Default output format [None]: 
+
+  $ aws --endpoint-url http://127.1:9000 s3 mb s3://irods
```
[the remark this answers is not in the document]
Comment (trel) [reply MISC.rst:29]: it's valid-ish (missings are assumed to be zero).  but not good.  please use the full 127.0.0.1

File: MISC.rst (hunk 2 above)
Comment (trel) [reply MISC.rst:1]: well, exactly... i think there may be a good cause for breaking things into a book or chapters or pages...   but i'd like to understand how pypi deals with that... can they be linked back together, etc...   before we do it.

yes just add a section at the bottom of the main README please

File: MISC.rst (hunk 9 above)
[the remark this answers is not in the document]
Comment (d-w-moore) [reply MISC.rst:29]: omg how did that get in there?

File: irods/manager/data_object_manager.py [hunk 10]
```diff
@@ -68,6 +71,7 @@ def should_parallelize_transfer( self,
             if not isinstance(size,six.integer_types):
                 size = obj_sz.tell()
             obj_sz.seek(pos,os.SEEK_SET)
+            if isinstance(measured_obj_size,list): measured_obj_size[:] = [size]
```
[the remark this answers is not in the document]
Comment (d-w-moore) [reply irods/manager/data_object_manager.py:74]: `measured_obj_size` then "becomes" [size] without going through a change of id (allocation/deallocation).  This is important when it's been passed by value (as are all Python function params) into the function, to be used as an output variable.
```
>>> def f(x): x=[3]
... 
>>> L=[2]; f(L)
>>> L
[2]
>>> def g(x): x[:]=[3]
... 
>>> g(L)
>>> L
[3]

```

File: irods/manager/data_object_manager.py (hunk 5 above)
Comment (d-w-moore) [reply irods/manager/data_object_manager.py:135]: Yes, it's passed as the obj_sz.  (This function is written so that a file or integer-like entity can be placed in that parameter - I forget exactly why though...)

File: irods/parallel.py (hunk 7 above)
Comment (d-w-moore) [reply irods/parallel.py:313]: Yep , it'll be the data object.  Will make it more descriptive.

File: irods/parallel.py (hunk 8 above)
Comment (d-w-moore) [reply irods/parallel.py:355]: Will do.

File: irods/parallel.py (hunk 6 above)
Comment (d-w-moore) [reply irods/parallel.py:324]: yes, it's the multipart size, the #bytes transferred per thread.  Maybe I'll use `partial_size`;

**Comment by d-w-moore** [comment]
> need more documentation in the code about 'why' and 'reasoning'. lots of very subtle changes here.

@trel are we in favor of more general comments throughout, to explain rationale behind all of the  parallel transfer
implementation's various parts, or is the new explanation needed in reference to **only** S3 related changes?

There is a lot of referendum and questions on old code in this review, so maybe it's more the former that is needed. Thoughts?

**Comment by trel** [comment]
In general, we want to write code that provides not just a functional interface or algorithm, but also teaches whoever is reading the code about what it's doing and why... but also about the frameworks or mental models helpful to understand how it fits in with the iRODS ecosystem.

We already know that the go-irodsclient is modeled after this library.

So, I have no specific requests, just that put on the shoes of the next person tasked with understanding what's being done in this library.  Start with S3, since it's so new, but yes, all the parallel code eventually... can be over time...

**Comment by d-w-moore** [comment]
That's what I was thinking ... while I'm dealing with it directly, I'll elucidate with better variable names per Kory's request, and put in skeletal comments to show what various code sections are doing. And give some insight particularly on the new S3 related things, if needed.

File: README.rst [hunk 11]
```diff
@@ -1086,3 +1086,83 @@ And more...
 -----------
 
 Additional code samples are available in the `test directory <https://github.com/irods/python-irodsclient/tree/master/irods/test>`_
+
+
+=======
+Testing
+=======
+
+Setting up and runnning tests
```
Comment (trel) [inline README.rst:1095]: runnning -> running

File: README.rst [hunk 12]
```diff
@@ -1086,3 +1086,83 @@ And more...
...
+-----------------------------
+
+The Python iRODS Client comes with its own suite of tests.  Some amount of setting up may be necessary first:
+
+    1. Use :code:iinit to specify the iRODS client environment.
```
Comment (trel) [inline README.rst:1100]: add backticks around iinit

File: README.rst [hunk 13]
```diff
@@ -1086,3 +1086,83 @@ And more...
...
+       For best results, point the client at a server running on the local host.
+
+    2. Install the python-irodsclient along with the :code:`unittest unittest_xml_reporting` module or the older :code:`xmlrunner` equivalent.
+          - for PRC versions 1.1.1 and later: :code:`pip install ./path-to-python-irodsclient-repo[tests]`  (when using a local Git repo)
+                                          or: :code:`pip install python-irods[tests]`  (when installing directly from PyPI)
```
Comment (trel) [inline README.rst:1105]: see how these render at github by looking at your branch's README.   i suggest doing...

later:
 :code:
or:
 :code:

to break up the words from the code itself, easier to read.

File: README.rst [hunk 14]
```diff
@@ -1086,3 +1086,83 @@ And more...
...
+          - earlier releases (<= 1.1.0) will install the outmoded :code:`xmlrunner` module automatically
+
+    3. Follow further instructions in the `test README <https://github.com/irods/python-irodsclient/tree/master/irods/test/README.rst>`_
```
Comment (trel) [inline README.rst:1108]: let's make it the same as line 1088

> Additional code samples are available in the `test directory <https://github.com/irods/python-irodsclient/tree/master/irods/test>`_

File: README.rst [hunk 15]
```diff
@@ -1086,3 +1086,83 @@ And more...
...
+Testing S3 parallel transfer
+----------------------------
+
+System requirements::
+
+- Ubuntu 18 user with Docker installed.
+- Local instance of iRODS server running.
+- Logged in sudo privileges.
+
+Run a MINIO service::
```
Comment (trel) [inline README.rst:1120]: MINIO -> MinIO

File: README.rst [hunk 16]
```diff
@@ -1086,3 +1086,83 @@ And more...
...
+
+  $ docker run -d -p 9000:9000 -p 9001:9001 minio/minio server /data --console-address ":9001" 
+
+Set up a bucket :code:`s3://irods` under MINIO::
```
Comment (trel) [inline README.rst:1124]: MINIO -> MinIO

File: README.rst [hunk 17]
```diff
@@ -1086,3 +1086,83 @@ And more...
...
+
+  $ pip install awscli
+
+  $ aws configure 
+  AWS Access Key ID [None]: minioadmin
+  AWS Secret Access Key [None]: minioadmin
+  Default region name [None]: 
+  Default output format [None]: 
+
+  $ aws --endpoint-url http://127.1:9000 s3 mb s3://irods
```
Comment (trel) [inline README.rst:1134]: 127.0.0.1 please

Comment (d-w-moore) [reply README.rst]: check.  This was pasted straight from my command history, usually a bad idea : )

File: README.rst (hunk 16 above)
Comment (d-w-moore) [reply README.rst]: yep

File: README.rst (hunk 15 above)
Comment (d-w-moore) [reply README.rst]: done

File: README.rst (hunk 12 above)
Comment (d-w-moore) [reply README.rst]: done

File: README.rst (hunk 11 above)
Comment (d-w-moore) [reply README.rst]: corrected.

File: irods/parallel.py (hunk 6 above)
Comment (korydraughn) [reply irods/parallel.py:324]: or `bytes_per_thread`

Comment (d-w-moore) [reply irods/parallel.py:324]: even better!

File: irods/manager/data_object_manager.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (d-w-moore) [reply irods/manager/data_object_manager.py:135]: A search revealed these:

https://stackoverflow.com/questions/4702249/is-there-a-way-in-python-to-return-a-value-via-an-output-parameter
https://docs.python.org/3/faq/programming.html#how-do-i-write-a-function-with-output-parameters-call-by-reference

I don't believe there's anything along the lines of a decorator or annotation for this, and when you get down to it, and - simply put - the Python list is the best way to do it.  Some offer that returning tuples is a better way, but when the output parameter _may or may not_ be calcuated and returned, None becomes a possible second value of the tuple to be returned. In my view, this would harm the simple semantics of `should_parallelize...` returning a boolean, and muddle the logic of  the calling code even worse.

[the remark this answers is not in the document]
Comment (d-w-moore) [reply irods/manager/data_object_manager.py:135]: Resolving for now.

File: irods/manager/data_object_manager.py [hunk 18]
```diff
@@ -110,7 +109,8 @@ def _download(self, obj, local_path, num_threads, **options):
 
 
     def get(self, path, local_path = None, num_threads = DEFAULT_NUMBER_OF_THREADS, **options):
-        """Get a reference to the data object at the specified `path'.
+        """
+        Get a reference to the data object at the specified `path'.
```
Comment (korydraughn) [inline irods/manager/data_object_manager.py:113]: is the character after `path` supposed to be a backtick or single quote?

**Review by korydraughn — COMMENTED** [review]
being able to change args on the `deferred_call` type is cool stuff.

File: README.rst [hunk 19]
```diff
@@ -1107,3 +1107,87 @@ And more...
...
+   For best results, point the client at a server running on the local host.
+
+2. Install the python-irodsclient along with the :code:`unittest unittest_xml_reporting` module or the older :code:`xmlrunner` equivalent.
+
+   - for PRC versions 1.1.1 and later:
+
+     *  :code:`pip install ./path-to-python-irodsclient-repo[tests]`  (when using a local Git repo); or,
+     *  :code:`pip install python-irodsclient[tests]'>=1.1.1'`  (when installing directly from PyPI).
+
+   - earlier releases (<= 1.1.0) will install the outmoded :code:`xmlrunner` module automatically
```
[the remark this answers is not in the document]
Comment (trel) [reply README.rst:1131]: not wrong... but consider ... 'outdated'?

File: irods/manager/data_object_manager.py (hunk 18 above)
Comment (d-w-moore) [reply irods/manager/data_object_manager.py:113]: It's a stylized thing sometimes used to set off a word visibly.  I should probably leave out both "quote" characters, esp if it tends to cause distraction or double-take... @trel?

File: README.rst (hunk 19 above)
[the remark this answers is not in the document]
Comment (d-w-moore) [reply README.rst:1131]: agreed

**Comment by d-w-moore** [comment]
> being able to change args on the `deferred_call` type is cool stuff.

Indeed, it was useful for this case too.

File: irods/manager/data_object_manager.py (hunk 18 above)
Comment (korydraughn) [reply irods/manager/data_object_manager.py:113]: you're saying mixing them is on purpose?

what i see is `<backtick>path<single_quote>`.

Comment (d-w-moore) [reply irods/manager/data_object_manager.py:113]: Yes, it was on purpose.  It seemed appropriate but may not be :)

Comment (korydraughn) [reply irods/manager/data_object_manager.py:113]: Understood. I think it matters for the next developer more than someone like me. If other developers understand it then it is likely ok.

Is this specific to python?

Comment (trel) [reply irods/manager/data_object_manager.py:113]: seems fine if it renders cleanly wherever the docstrings get rendered...

Comment (d-w-moore) [reply irods/manager/data_object_manager.py:113]: Not particularly Python-related, no. I've seen it in various spots around the internet, sometimes in documentation, probably more to do with style than accuracy.

Comment (d-w-moore) [reply irods/manager/data_object_manager.py:113]: @trel yes it seems to render fine ....

**Comment by d-w-moore** [comment]
squashed but not yet #'d

**Comment by trel** [comment]
consider rewriting the first commit message to be more focused.

File: README.rst [hunk 20]
```diff
@@ -1107,3 +1107,87 @@ And more...
...
+2. Install the python-irodsclient along with the :code:`unittest unittest_xml_reporting` module or the older :code:`xmlrunner` equivalent.
+
+   - for PRC versions 1.1.1 and later:
+
+     *  :code:`pip install ./path-to-python-irodsclient-repo[tests]`  (when using a local Git repo); or,
+     *  :code:`pip install python-irodsclient[tests]'>=1.1.1'`  (when installing directly from PyPI).
+
+   - earlier releases (<= 1.1.0) will install the outdated :code:`xmlrunner` module automatically
+
+3. Follow further instructions in the `test directory <https://github.com/irods/python-irodsclient/tree/master/irods/test>`_
```
Comment (trel) [inline README.rst:1133]: master -> main

**Comment by d-w-moore** [comment]
More detail, or just a better headline ?

**Comment by trel** [comment]
> More detail, or just a better headline ?

it's currently multiline, but looks just like a squashing of multiple earlier one line commit messages.

File: README.rst (hunk 20 above)
Comment (d-w-moore) [reply README.rst]: changed both links that were pointing to a .../master/... link to  .../main/... instead.

**Comment by d-w-moore** [comment]
> > More detail, or just a better headline ?
> 
> it's currently multiline, but looks just like a squashing of multiple earlier one line commit messages.

prettied it up.

**Comment by trel** [comment]
yeps, squash it

**Comment by d-w-moore** [comment]
Squashed to minimum commits.

**Comment by d-w-moore** [comment]
Ok, squashed to minimum now.

**Comment by trel** [comment]
awesome - please # them.

**Comment by d-w-moore** [comment]
#'d
